wal/src/writer_thread.rs

```diff
@@ -0,0 +1,151 @@
+            }),
+        }
+    }
+
+    /// Enqueue `batch` to be wrote to the current open segment file.
+    ///
+    /// Once flushed to disk and durable (or the write failed) the write result
+    /// is broadcast through the embedded channel.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the I/O thread is not running.
+    pub(crate) async fn enqueue_batch(&self, batch: WalBuffer) {
+        self.inner
+            .as_ref()
+            .unwrap()
+            .batch_tx
+            .send(batch)
+            .await
+            .expect("wal writer IO thread is dead")
+    }
+}
+
+impl Drop for WriterIoThreadHandle {
+    fn drop(&mut self) {
+        let inner = self.inner.take().unwrap();
+        // Signal to the IO thread that it should stop.
+        drop(inner.batch_tx);
+        // Wait for the IO thread to gracefully stop, discarding any errors -
+        // attempting to unwrap/panic in this Drop implwill cause an immediate
+        // SIGSEV.
+        let _ = inner.join_handle.join();
+    }
+}
+
+/// The state of the I/O actor thread.
+struct WriterIoThread {
+    /// A channel to receive batches to flush.
+    batch_rx: mpsc::Receiver<WalBuffer>,
+    /// The set of segments, used to obtain the current open segment handle.
+    segments: Arc<Mutex<Segments>>,
+}
+
+impl WriterIoThread {
+    fn new(batch_rx: mpsc::Receiver<WalBuffer>, segments: Arc<Mutex<Segments>>) -> Self {
+        Self { batch_rx, segments }
+    }
+
+    fn run(mut self) {
+        // Maintain a serialisation buffer for re-use, instead of allocating a
+        // new one for each batch.
+        let mut proto_data = Vec::with_capacity(4 * 1024);
+
+        loop {
+            proto_data.clear();
+
+            let batch = match self.batch_rx.blocking_recv() {
+                Some(batch) => batch,
+                None => {
+                    // The batch channel has closed - all handles have been
+                    // dropped.
+                    debug!("stopping WAL IO thread");
+                    return;
+                }
+            };
+
+            // Encode the batch into the proto types
+            let ops: Vec<_> = batch
+                .ops
+                .into_iter()
+                .map(proto::SequencedWalOp::from)
+                .collect();
+            let proto_batch = proto::WalOpBatch { ops };
+
+            // Generate the binary protobuf message, storing it into proto_data
+            proto_batch
+                .encode(&mut proto_data)
+                .expect("encoding batch into vec cannot fail");
+
+            // Write the serialised data to the current open segment file.
+            let res = {
+                let mut segments = self.segments.lock();
+                match segments.open_segment.write(&proto_data) {
+                    Ok(summary) => WriteResult::Ok(summary),
+                    Err(e) => {
+                        warn!(erorr=%e, "failed to write WAL batch");
```

Comment:
Typo that might affect searching logs:

```suggestion
                        warn!(error=%e, "failed to write WAL batch");
```

Reply:
🤦